### PR TITLE
Add initial LOBPCG top-k eigenvalue solver (#3112)

### PR DESCRIPTION
**Pull request by vlad17**
This initial version is f32-only for accelerators, since it relies on an eigh call (which itself is f32 at most) in its inner loop.

For details, see jax.experimental.linalg.standard_lobpcg documentation.

### CODE REVIEW

**Comment by vlad17** [comment]
@shoyer @tabakg @lobpcg would any of you be interested in reviewing?

File: tests/lobpcg_test.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+class LobpcgTest(jtu.JaxTestCase):
+  # Override base class default config.
+  _default_config = {
+      'jax_enable_checks': True,
+      'jax_debug_nans': True,
+      'jax_numpy_rank_promotion': 'raise',
+      'jax_traceback_filtering': 'off',
+  }
+
+  # TODO(vladf): add f64 tests just to verify it compiles?
```
Comment (tabakg) [inline tests/lobpcg_test.py:102]: I think that would be good -- you could probably check if it's running on CPU and disable otherwise?

**Comment by shoyer** [comment]
A couple high level thoughts:

1. It would be nice to support matrix-free linear operators defined on pytrees, like the solvers in jax.scipy.sparse.linalg. For example, imagine solving for the largest eigenvalue pairs of a function that evaluates a linearized neural network.
2. It would be nice to support an iterative interface with init and update functions, similar to the optimizers in Optax. This sort of inversion of control provides valuable flexibility in cases where a full eigenvalue solve from scratch is prohibitively expensive. For example, imagine performing a single LOBPCG step to re-estimate the top eigenvalue after each gradient descent step when training a neural net, as is done in spectral normalization.

Neither of these are deal breakers for the first iteration, though.

**Comment by vlad17** [comment]
> It would be nice to support matrix-free linear operators defined on pytrees, like the solvers in jax.scipy.sparse.linalg. For example, imagine solving for the largest eigenvalue pairs of a function that evaluates a linearized neural network.

Great idea, this should be easy to add and now that I think about it, and I think Shankar (skrishnan@google.com) was another user who'd immediately benefit from an interface like that.

> It would be nice to support an iterative interface with init and update functions, similar to the optimizers in Optax. This sort of inversion of control provides valuable flexibility in cases where a full eigenvalue solve from scratch is prohibitively expensive. For example, imagine performing a single LOBPCG step to re-estimate the top eigenvalue after each gradient descent step when training a neural net, as is done in spectral normalization.

Could you elaborate on this? I'd be very willing to do this as a follow-on if I had a user to work with on their ideal API for this, but as-stated I don't quite see how setting the initial `X` to the previous value and maximum iteration `m=1` wouldn't work.

**Comment by shoyer** [comment]
> > It would be nice to support an iterative interface with init and update functions, similar to the optimizers in Optax. This sort of inversion of control provides valuable flexibility in cases where a full eigenvalue solve from scratch is prohibitively expensive. For example, imagine performing a single LOBPCG step to re-estimate the top eigenvalue after each gradient descent step when training a neural net, as is done in spectral normalization.
> 
> Could you elaborate on this? I'd be very willing to do this as a follow-on if I had a user to work with on their ideal API for this, but as-stated I don't quite see how setting the initial `X` to the previous value and maximum iteration `m=1` wouldn't work.

Wouldn't we want to calculate the matrix P from the previous iteration?

**Comment by vlad17** [comment]
Ah, gotcha. I guess that makes sense, and it seems like it'd be a matter of exposing the `body` function in the current method. Maybe it'd be best to wait until the API solidifies a little though, since `largest=False` option is incoming, and it'd be good to figure out how that iterative interface would work with a preconditioner.

File: tests/lobpcg_test.py (hunk 1 above)
Comment (vlad17) [reply tests/lobpcg_test.py:98]: It looks like there are `f64` tests auto-triggered by the github action matrix, so I think all I should  really need to do here is adjust the epsilons.

**Comment by vlad17** [comment]
@shoyer @tabakg just updated the pr on a flight home, let me know what you think of the new interface and tests.

**Comment by vlad17** [comment]
Here are some cool curves (which can be generated from the tests) of jax vs scipy f32 on 1000x1000 versions of the test matrices for top-10 eigs:

linear, geom, and ring laplacian

clustered eigenvals

**Comment by vlad17** [comment]
@lobpcg I used 1.8.0. The examples are all the same as the unit tests in the PR but size 1000.

The actual colab to make the viz depends on some Google-internal features for unrelated things, but I can try to find some time to clean up the notebook for a public-facing version. Would it be more appropriate to post that as a scipy github issue?

**Comment by lobpcg** [comment]
> @lobpcg I used 1.8.0 (did anything change in 1.8.1?). The examples are all the same as the unit tests in the PR but size 1000.
> 
> The actual colab to make the viz depends on some Google-internal features for unrelated things, but I can try to find some time to clean up the notebook for a public-facing version. Would it be more appropriate to post that as a scipy github issue? Just to keep this thread focussed on jax.

1.8.0 is representative. I made changes in lobpcg specifically for float32, but that was before.

Of course if you create the reproducible issue in SciPy that would be ideal. No need to include the code to make the actual  plots, just please add a reference to your post with the plots above. 

If SciPy fails on smaller sizes like 100, all those would be good examples to include into SciPy as unit tests, if I find a fix so that they all run.

**Comment by vlad17** [comment]
@lobpcg filled out https://github.com/scipy/scipy/issues/16408 with just the cases, no viz.

File: jax/experimental/sparse/linalg.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+    tol: Union[jnp.ndarray, float, None] = None):
+  """Compute the top-k standard eigenvalues using the LOBPCG routine.
+
+  LOBPCG [1] stands for Locally Optimal Block Preconditioned Conjugate Gradient.
+  The method enables finding top-k eigenvectors in an accelerator-friendly
+  manner.
+
+  This initial experimental version has several caveats.
+
+    - Only the standard eigenvalue problem `A U = lambda U` is supported,
```
Comment (rmlarsen) [inline jax/experimental/sparse/linalg.py:38]: This MR seems a little odd to me. The power of LOBPCG is primarily on finding the smallest eigenvalues when a preconditioner for A is known. For finding the largest, the (block) Lanczos algorithm is provably optimal and well-understood.

File: jax/experimental/sparse/linalg.py [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+  # such that (2) V is (S.T S)-orthonormal.
+  # https://www.netlib.org/lapack/lug/node54.html
+  #
+  # Usually it requires solving the complicated standard eigensystem
+  # U^-T S^T A S U^-1 @ Q = w * Q and then backsolving V = U^-1 Q,
+  # but if S is standard orthonormal then we just need to find
+  # eigenvalues of S.T A S.
+
+  SAS = _mm(S.T, A(S))
+  SAS = (SAS + SAS.T) / 2
```
Comment (rmlarsen) [inline jax/experimental/sparse/linalg.py:364]: Explain why we need to symmetrize. I've seen cases where it is actually detrimental to accuracy (assuming the eigensolver only uses one triangle of the input. Did you try to do it without?

File: jax/experimental/sparse/linalg.py [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+
+
+def _extend_basis(X, m):
+  # Use a block householder reflector to generate orthogonal extension
+  # to X. There's nothing too special about this, and we could choose
+  # any random extension to X's basis, but this is a deterministic choice.
+  #
+  # https://epubs.siam.org/doi/abs/10.1137/0725014
+  # https://www.jstage.jst.go.jp/article/ipsjdc/2/0/2_0_298/_article
+  n, k = X.shape
```
Comment (rmlarsen) [inline jax/experimental/sparse/linalg.py:378]: Please write out the math here instead of just citing the papers. The code is not particularly readable with all the concatenations etc.

**Review by rmlarsen — CHANGES_REQUESTED** [review]
I

File: jax/experimental/sparse/linalg.py (hunk 2 above)
Comment (vlad17) [reply jax/experimental/sparse/linalg.py:38]: Discussed offline: this PR adds just top-k to make the first change as small as we can get.

The extra preconditioner param and support for k-smallest eigenvals will come in a follow up PR(s) which will help reduce reviewer load for testing/locs etc.

File: jax/experimental/sparse/linalg.py (hunk 4 above)
Comment (vlad17) [reply jax/experimental/sparse/linalg.py:473]: done

File: jax/experimental/sparse/linalg.py [hunk 5]
```diff
@@ -291,20 +294,51 @@ def _eigh_ascending(A):
...
-  # a degenerate, non-orthogonal one.
+  """Derives a truncated orthonormal basis for `X`.
 
+  SVQB [1] is an accelerator-friendly orthonormalization procedure, which
+  squares the matrix `C = X.T @ X` and computes an eigenbasis for a smaller
+  `(k, k)` system; this offloads most of the work in orthonormalization
+  to the first multiply when `n` is large.
+
+  Importantly, if diagonalizing the squared matrix `C` reveals numerical
+  degeneracy in its spectrum (which would be evidenced by near-0 eigenvalues),
```
Comment (rmlarsen) [inline jax/experimental/sparse/linalg.py:305]: s/degeneracy in its spectrum/rank deficiency of X/

**Review by rmlarsen — APPROVED** [review]

File: jax/experimental/sparse/linalg.py [hunk 6]
```diff
@@ -291,20 +294,51 @@ def _eigh_ascending(A):
 
 
 def _svqb(X):
-  # https://sdm.lbl.gov/~kewu/ps/45577.html
-  # May choose to truncate the basis rather than return
-  # a degenerate, non-orthogonal one.
+  """Derives a truncated orthonormal basis for `X`.
 
+  SVQB [1] is an accelerator-friendly orthonormalization procedure, which
+  squares the matrix `C = X.T @ X` and computes an eigenbasis for a smaller
```
[the remark this answers is not in the document]
Comment (rmlarsen) [reply jax/experimental/sparse/linalg.py:300]: Rank revealing QR would be a nice options if it was available in JAX. It might be worth adding. A version based on the randomized projection approach could be made accelerator-friendly.

[the remark this answers is not in the document]
Comment (vlad17) [reply jax/experimental/sparse/linalg.py:300]: By QR here, you mean QR-ing `X` itself not its square right? At one point early in the algorithm development process I tried that but decided against it b/c the speed wasn't satisfactory. Cholesky on XTX would be faster but rank deficiency is specifically a case that needs to be handled here.

That said, I'll put trying the QR approach out again on my TODO list. It's worth revisiting and I don't have hard numbers saying it's too slow. And it avoids squaring.

I'm really intrigued by the randomized projection you're mentioning. Do you have a reference?

File: jax/experimental/sparse/linalg.py [hunk 7]
```diff
@@ -291,20 +294,51 @@ def _eigh_ascending(A):
...
 
   inner = _mm(X.T, X)
 
   w, V = _eigh_ascending(inner)
 
+  # All mask logic is used to avoid divide-by-zeros when input columns
+  # may have been zero or new zero columns introduced from truncation.
+  #
+  # If an eigenvalue is less than max eigvalue * eps, then consider
+  # that direction degenerate.
```
[the remark this answers is not in the document]
Comment (vlad17) [reply jax/experimental/sparse/linalg.py:333]: My usage here is bleeding in from optimization where you'd refer to non-full-rank constraints as a degenerate problem. Regardless, I put it in quotes.

File: jax/experimental/sparse/linalg.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rmlarsen) [reply jax/experimental/sparse/linalg.py:300]: Here is a reference for the RRQR with randomized projection: https://arxiv.org/abs/2008.04447

You are right that computing the QR decomposition of X is likely slower, since `X.T@X` is so fast on an accelerator. But avoiding the squaring would be nice. Just some things to experiment with, I guess.

File: jax/experimental/sparse/linalg.py [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+
+def _project_out(basis, U):
+  # "twice is enough" from shoyer's reference:
+  # http://slepc.upv.es/documentation/reports/str1.pdf
+
+  for _ in range(2):
+    U -= _mm(basis, _mm(basis.T, U))
+    U = _orthonormalize(U)
+
+  # It's crucial to end on a subtraction of the original basis.
```
[the remark this answers is not in the document]
Comment (rmlarsen) [reply jax/experimental/sparse/linalg.py:395]: Hmm this is very interesting. If you allow these small non-zero components in the complement of the span of the large column to be normalized, it can act as a kind on randomized restarting, but I agree that it is difficult to control. I used it in Lanczos (PROPACK) to recover from computing an invariant subspace before convergence, but it is tricky.

[the remark this answers is not in the document]
Comment (vlad17) [reply jax/experimental/sparse/linalg.py:395]: Indeed! I hope to find time to at least dig into it to do an internal presentation on it, it seemed like a nice new learning from implementing this.

One approach that also worked, but had slightly worse quality, was to leave these "random restarts" in and do another svqb() to orthonormalize the Ritz basis (this is also more expensive b/c the whole 3k x 3k system needs to be mutually orthonormalized).
